### ISSUE TITLE
Feedback on `cudatoolkit` packages

### ISSUE THREAD

**Issue by jakirkham** [conda-forge/cudatoolkit-feedstock#15]
Opening to collect feedback on `cudatoolkit` packages. The packages are currently under the `test` label. So installing requires doing `conda install -c conda-forge/label/test cudatoolkit`. Would be good if people tried them out and reported back any feedback or issues encountered. Copying a few people who may be interested below (please feel free to add others).

cc @leofang @jaimergp @gmarkall @ax3l @pearu @carterbox @cjnolet @dantegd @teju85 @raydouglass @mike-wendt @kkraus14 @quasiben

**Comment by pearu** [comment]
IIUC, NVIDIA EULA allows only download and use the cudatoolkit (if one accepts the license) by the end-user, however, other conda-forge packages cannot depend on it because these would use the cudatoolkit in CI instances when building or testing the packages (if conda-forge CI instances would have GPU devices, perhaps in future). Is my understanding of the limitations of using this package correct?

Also, how this package compares to https://github.com/conda-forge/cudatoolkit-dev-feedstock ?

**Comment by scopatz** [comment]
The CI is the end user in that context @pearu. It isn't being redistributed from the CI.

cudatoolkit-dev installs the whole of cudatoolkit in a post-link script.  It is probably a bad idea and that repo should be archived once a transition here is done.

**Comment by jakirkham** [comment]
Just to add to what Anthony said, we did a few things to make this package acceptable to stakeholders at NVIDIA.

First the package has [a post-link script]( https://github.com/conda-forge/cudatoolkit-feedstock/blob/d2183045f96cbd679be5f9d47172d62775c93168/recipe/post-link.sh ) that notes the user accepts the EULA when it is installed. Second we went through some effort to include relevant info on [the Anaconda.org `cudatoolkit` package page]( https://anaconda.org/conda-forge/cudatoolkit ) as noted in issue ( https://github.com/conda-forge/cudatoolkit-feedstock/issues/2 ). Third this package went through an internal review process at NVIDIA, which is tracked by an internal link referenced in the recipe ( https://github.com/conda-forge/cudatoolkit-feedstock/pull/10 ).

**Comment by carterbox** [comment]
My understanding is that this package should be a drop-in replacement for the `main/cudatoolkit` package, so I tested by installing the current CUDA-enabled build of `tomopy` with `conda-forge/label/test` having priority.

11 - conda will not choose `conda-forge/label/test/cudatoolkit` it always chooses from `main` instead.
10, 9 -  `conda-forge/label/test/cudatoolkit` is missing libnppicom.so.10 and its aliases, so `tomopy` fails to import.

**Comment by jaimergp** [comment]
On Linux 64, it defaults to cudatoolkit=10.2 instead of 11 if now `cudatoolkit` version is specified. I guess that's because `cudatoolkit=11` on linux-64 is [now labeled as `broken`](https://anaconda.org/conda-forge/cudatoolkit/files)?

```
% conda info | grep -e "cuda" -e "https"
virtual packages : __cuda=11.0
channel URLs : https://conda.anaconda.org/conda-forge/linux-64
               https://conda.anaconda.org/conda-forge/noarch
               https://repo.anaconda.com/pkgs/main/linux-64
               https://repo.anaconda.com/pkgs/main/noarch
               https://repo.anaconda.com/pkgs/r/linux-64
               https://repo.anaconda.com/pkgs/r/noarch
               https://conda.anaconda.org/omnia/linux-64
               https://conda.anaconda.org/omnia/noarch

% conda create -yq -n openmm-cf -c conda-forge/label/testing -c conda-forge/label/test openmm

% conda activate openmm-cf

% conda list | grep cudatoolkit
cudatoolkit               10.2.89              he1b5a44_0    conda-forge/label/test

% python -m simtk.testInstallation
OpenMM Version: 7.4.2
Git Revision: 9e6677a66b6e4f32918f0635e7a72218d1abe277

There are 4 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 CUDA - Successfully computed forces
4 OpenCL - Successfully computed forces

Median difference in forces between platforms:

Reference vs. CPU: 6.29039e-06
Reference vs. CUDA: 6.73097e-06
CPU vs. CUDA: 7.81523e-07
Reference vs. OpenCL: 6.75312e-06
CPU vs. OpenCL: 8.17069e-07
CUDA vs. OpenCL: 2.20998e-07

All differences are within tolerance.
```

If I force cudatoolkit versions:

11.0 🔴 (comes from defaults)


```
% conda create -yq -n openmm-cf -c conda-forge/label/testing -c conda-forge/label/test openmm cudatoolkit=11

% conda activate openmm-cf

% conda list --show-channel-urls | grep cudatoolkit
cudatoolkit               11.0.221             h6bb024c_0    defaults

% python -m simtk.testInstallation

OpenMM Version: 7.4.2
Git Revision: 9e6677a66b6e4f32918f0635e7a72218d1abe277

There are 4 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 CUDA - Successfully computed forces
4 OpenCL - Successfully computed forces

Median difference in forces between platforms:

Reference vs. CPU: 6.30387e-06
Reference vs. CUDA: 6.72922e-06
CPU vs. CUDA: 7.79891e-07
Reference vs. OpenCL: 6.75312e-06
CPU vs. OpenCL: 8.15089e-07
CUDA vs. OpenCL: 2.19328e-07

All differences are within tolerance.
```

10.1 🟢 

```
% conda create -yq -n openmm-cf -c conda-forge/label/testing -c conda-forge/label/test openmm cudatoolkit=10.1
conda activate openmm-cf

% conda list --show-channel-urls | grep cudatoolkit
cudatoolkit               10.1.243             he1b5a44_0    conda-forge/label/test

% python -m simtk.testInstallation

OpenMM Version: 7.4.2
Git Revision: 9e6677a66b6e4f32918f0635e7a72218d1abe277

There are 4 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 CUDA - Successfully computed forces
4 OpenCL - Successfully computed forces

Median difference in forces between platforms:

Reference vs. CPU: 6.30122e-06
Reference vs. CUDA: 6.73932e-06
CPU vs. CUDA: 7.77778e-07
Reference vs. OpenCL: 6.75312e-06
CPU vs. OpenCL: 8.15305e-07
CUDA vs. OpenCL: 2.20438e-07

All differences are within tolerance.
```

10.0 🟢 

```
% conda create -yq -n openmm-cf -c conda-forge/label/testing -c conda-forge/label/test openmm cudatoolkit=10.0
% conda activate openmm-cf
% conda list --show-channel-urls | grep cudatoolkit
cudatoolkit               10.0.130             he1b5a44_0    conda-forge/label/test
% python -m simtk.testInstallation

OpenMM Version: 7.4.2
Git Revision: 9e6677a66b6e4f32918f0635e7a72218d1abe277

There are 4 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 CUDA - Successfully computed forces
4 OpenCL - Successfully computed forces

Median difference in forces between platforms:

Reference vs. CPU: 6.30118e-06
Reference vs. CUDA: 6.72919e-06
CPU vs. CUDA: 7.75283e-07
Reference vs. OpenCL: 6.75312e-06
CPU vs. OpenCL: 8.09844e-07
CUDA vs. OpenCL: 2.19749e-07

All differences are within tolerance.
```

9.2 🟢 

```
% conda create -yq -n openmm-cf -c conda-forge/label/testing -c conda-forge/label/test openmm cudatoolkit=9.2
% conda activate openmm-cf
% conda list --show-channel-urls | grep cudatoolkit
cudatoolkit               9.2.148              he1b5a44_0    conda-forge/label/test
% python -m simtk.testInstallation

OpenMM Version: 7.4.2
Git Revision: 9e6677a66b6e4f32918f0635e7a72218d1abe277

There are 4 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 CUDA - Successfully computed forces
4 OpenCL - Successfully computed forces

Median difference in forces between platforms:

Reference vs. CPU: 6.30149e-06
Reference vs. CUDA: 6.73832e-06
CPU vs. CUDA: 7.75635e-07
Reference vs. OpenCL: 6.75312e-06
CPU vs. OpenCL: 8.1356e-07
CUDA vs. OpenCL: 2.19948e-07

All differences are within tolerance.
```

**Comment by jaimergp** [comment]
When it comes to _building_ packages, I think we do have an issue. All `cudatoolkit` packages on conda-forge seem to depend on glibc>= 2.17, regardless the version.

![image](https://user-images.githubusercontent.com/2559438/97421299-a3013780-190c-11eb-84dc-6d857b5756a8.png)


AFAIK, CentOS 7 is only needed for CUDA 11+, so CUDA <= 10.2 should be built with glibc => 2.12, right?

**Comment by jaimergp** [comment]
The `cdt_name` issue shouldn't affect CUDA 11 build though, but these are not building anyway for other reasons. I think the label  `test` is too general and might have been used by other (broken) packages, which introduces solve conflicts. May I suggest using a more specific label like `cudatoolkit-test` so strict priority does not complicate things too much? Thanks!

**Comment by gmarkall** [comment]
On `linux-64` the toolkit versions 9.1, 10.0, 10.1, and 10.2 all seem to work with Numba - I didn't try with 11.0 as it's marked as broken at present. I've created environments with (example for 9.2):

```
conda create -c conda-forge -c conda-forge/label/test -n cfnumbacuda92 python=3.8 numba cudatoolkit=9.2
```

and tested with

```
python -c "from numba import cuda; cuda.test()"
```

**Comment by gmarkall** [comment]
I've also tested on `win-64` now with 10.1, 10.2 and 11.0 - all Numba tests run fine on Windows with these packages too.

**Comment by scopatz** [comment]
> missing libnppicom.so and its aliases, so tomopy fails to import

@carterbox - thanks for pointing this out! PRs welcome. It should be as simple as adding this lib name to the `cuda_libs` list in build.py.

>  Is it because cudatoolkit=11 on linux-64 is now labeled as broken?

@jaimergp - I think that is because @jakirkham or @isuruf may have manually moved the first package to the broken label initially. I have moved it to the test label to be consistent. 

> When it comes to building packages, I think we do have an issue

Yeah, for posterity, this conversation is mostly taking place at https://github.com/conda-forge/nvcc-feedstock/issues/48

> all Numba tests run fine on Windows with these packages too.

@gmarkall - :tada: amazing!

**Comment by scopatz** [comment]
> CentOS 7 is only needed for CUDA 11+, so CUDA <= 10.2 should be built with glibc => 2.12, right? In other words, this line needs a selector for CUDA 11+, and a 2.12 counterpart for the other versions.

Yeah, I think you are right about that.

**Comment by jaimergp** [comment]
> I have moved it [cudatoolkit=11] to the test label to be consistent.

Yep, that fixed it! 


11.0 🟢 

```
% conda create -yq -n openmm-cf -c conda-forge/label/testing -c conda-forge/label/test openmm cudatoolkit=11
% conda activate openmm-cf
% conda list --show-channel-urls | grep cudatoolkit
cudatoolkit               11.0.3               he1b5a44_0    conda-forge/label/test
% python -m simtk.testInstallation
python -m simtk.testInstallation

OpenMM Version: 7.4.2
Git Revision: 9e6677a66b6e4f32918f0635e7a72218d1abe277

There are 4 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 CUDA - Successfully computed forces
4 OpenCL - Successfully computed forces

Median difference in forces between platforms:

Reference vs. CPU: 6.29789e-06
Reference vs. CUDA: 6.72685e-06
CPU vs. CUDA: 7.74669e-07
Reference vs. OpenCL: 6.75312e-06
CPU vs. OpenCL: 8.08295e-07
CUDA vs. OpenCL: 2.14905e-07

All differences are within tolerance.
```

**Comment by scopatz** [comment]
@jaimergp - for the sysroot issue, there are now PRs #16 #17 #18 #19

**Comment by jaimergp** [comment]
Awesome, thanks!

**Comment by jakirkham** [comment]
> > Is it because cudatoolkit=11 on linux-64 is now labeled as broken?
> 
> @jaimergp - I think that is because @jakirkham or @isuruf may have manually moved the first package to the broken label initially. I have moved it to the test label to be consistent.

Yeah though I think we need to readd through the admin process for that to work ( https://github.com/conda-forge/admin-requests/pull/152 ).

**Comment by jakirkham** [comment]
Are people able to get the `11.0.3` package ok? @jaimergp?

**Comment by jaimergp** [comment]
Seems to be working fine (Linux 64):

```
% conda create -yq -n openmm-cf -c conda-forge/label/testing -c conda-forge/label/test openmm

Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... done

## Package Plan ##

  environment location: /home/rodriguj/.conda/envs/openmm-cf

  added / updated specs:
    - openmm


The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu
  ca-certificates    conda-forge/linux-64::ca-certificates-2020.6.20-hecda079_0
  certifi            conda-forge/linux-64::certifi-2020.6.20-py39h079e4ff_2
  cudatoolkit        conda-forge/label/test/linux-64::cudatoolkit-11.0.3-he1b5a44_0
  fftw               conda-forge/linux-64::fftw-3.3.8-nompi_hdcdd268_1112
  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.35-h769bd43_9
  libblas            conda-forge/linux-64::libblas-3.9.0-2_openblas
  libcblas           conda-forge/linux-64::libcblas-3.9.0-2_openblas
  libffi             conda-forge/linux-64::libffi-3.2.1-he1b5a44_1007
  libgcc-ng          conda-forge/linux-64::libgcc-ng-9.3.0-h5dbcf3e_17
  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-9.3.0-he4bcb1c_17
  libgfortran5       conda-forge/linux-64::libgfortran5-9.3.0-he4bcb1c_17
  libgomp            conda-forge/linux-64::libgomp-9.3.0-h5dbcf3e_17
  liblapack          conda-forge/linux-64::liblapack-3.9.0-2_openblas
  libopenblas        conda-forge/linux-64::libopenblas-0.3.12-pthreads_h4812303_1
  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-9.3.0-h2ae2ef3_17
  ncurses            conda-forge/linux-64::ncurses-6.2-he1b5a44_2
  numpy              conda-forge/linux-64::numpy-1.19.2-py39h2bb7b6c_1
  ocl-icd            conda-forge/linux-64::ocl-icd-2.2.12-h516909a_1005
  ocl-icd-system     conda-forge/linux-64::ocl-icd-system-1.0.0-1
  openmm             conda-forge/label/testing/linux-64::openmm-7.4.2-py39h66fb953_3
  openssl            conda-forge/linux-64::openssl-1.1.1h-h516909a_0
  pip                conda-forge/noarch::pip-20.2.4-py_0
  python             conda-forge/linux-64::python-3.9.0-h2a148a8_4_cpython
  python_abi         conda-forge/linux-64::python_abi-3.9-1_cp39
  readline           conda-forge/linux-64::readline-8.0-he28a2e2_2
  setuptools         conda-forge/linux-64::setuptools-49.6.0-py39h079e4ff_2
  sqlite             conda-forge/linux-64::sqlite-3.33.0-h4cf870e_1
  tk                 conda-forge/linux-64::tk-8.6.10-hed695b0_1
  tzdata             conda-forge/noarch::tzdata-2020d-h516909a_0
  wheel              conda-forge/noarch::wheel-0.35.1-pyh9f0ad1d_0
  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_1
  zlib               conda-forge/linux-64::zlib-1.2.11-h516909a_1010


Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... b'By downloading and using the CUDA Toolkit conda packages, you accept the terms and conditions of the CUDA End User License Agreement (EULA): https://docs.nvidia.com/cuda/eula/index.html\n'
done
```

**Comment by jakirkham** [comment]
Thanks Jaime! That's great to see 😄

**Comment by leofang** [comment]
Randomly picked a few tests to test the new `cudatoolkit` with CuPy from Conda-Forge...Still ongoing...
- [x] CUDA 9.2 + CuPy v8.1 + Python 3.7
- [ ] CUDA 10.0 + CuPy v8.1 + Python 3.7
- [ ] CUDA 10.1 + CuPy v8.1 + Python 3.7
- [ ] CUDA 10.2 + CuPy v8.1 + Python 3.7
- [ ] CUDA 11.0 + CuPy v8.1 + Python 3.7

**Comment by jakirkham** [comment]
Thanks for all of the PRs @carterbox and @scopatz for reviewing! 😄 Could you please recheck once the new packages are out? 😉

**Comment by carterbox** [comment]
TomoPy runs with the following cudatoolkit builds from `-c conda-forge/label/test/`
- [ ] 11.0
- [ ] 10.2
- [ ] 10.0
- [ ] 9.2

**Comment by jakirkham** [comment]
Great, thanks for testing! 😄

**Comment by jakirkham** [comment]
So it seems like these packages are working well now. Should we try moving them into the `main` label? 🙂

cc @scopatz @kkraus14 @mike-wendt

**Comment by isuruf** [comment]
Some files are still missing compared to `defaults`. For eg: `lib/libnvjpeg.so*` is missing in 10.2.

**Comment by isuruf** [comment]
Also are we authorized to distribute static libraries like `./lib/libcudadevrt.a`? I thought we could only distribute shared libraries. `defaults` is not distributing this file.

**Comment by isuruf** [comment]
We also have `libcuinj64.so` and `libaccinj64.so` distributed that defaults doesn't and is not mentioned in https://docs.nvidia.com/cuda/archive/10.2/eula/index.html#cuda-toolkit-supplement-license-agreement

**Comment by jakirkham** [comment]
> Some files are still missing compared to `defaults`. For eg: `lib/libnvjpeg.so*` is missing in 10.2.

AFAICT that should already be included ( https://github.com/conda-forge/cudatoolkit-feedstock/commit/a307b6c62708972cb35ab2de48a0d4c65d2a14d9 ).

> Also why are we distributing `./lib/libcudadevrt.a`? There's no need to have a static library at runtime. `defaults` is not distributing this file.

This was added to the Anaconda package in PR ( https://github.com/AnacondaRecipes/cudatoolkit-feedstock/pull/5 ). So has been included here as well. In the future this may be used by Numba ( https://github.com/numba/numba/pull/4551 ).

cc @gmarkall (who may have more thoughts here)

> We also have `libcuinj64.so` and `libaccinj64.so` distributed that defaults doesn't and is not mentioned in https://docs.nvidia.com/cuda/archive/10.2/eula/index.html#cuda-toolkit-supplement-license-agreement

I'm not really sure about that. AFAICT we are not including these libraries in [our list]( https://github.com/conda-forge/cudatoolkit-feedstock/blob/e9b3132f085307cf969b03e92aff74edd3aa1109/recipe/build.py#L78-L101 ).

Looking at `defaults` package they actually have these too. Not sure why that is either as they don't seem to do anything special to include them either.

The only mention of them I see in [the docs is in relation to NVTX]( https://docs.nvidia.com/cuda/profiler-users-guide/index.html#profiler-known-issues ).

@kkraus14 @mike-wendt, would either of you be able to comment here?

**Comment by isuruf** [comment]
> AFAICT that should already be included ( a307b6c ).

Sorry, I was looking at `10.1.243` and it isn't included compared to defaults.

> This was added to the Anaconda package in PR ( AnacondaRecipes/cudatoolkit-feedstock#5 ). So has been included here as well. 

static libraries should be in a separate package according to CFEP-18. 

> Looking at defaults package they actually have these too.

They aren't in `cudatoolkit-10.1.243-h6bb024c_0`

**Comment by gmarkall** [comment]
> Also why are we distributing `./lib/libcudadevrt.a`? There's no need to have a static library at runtime. `defaults` is not distributing this file.

This is needed for JIT compilers such as Numba, which compile and link code at runtime - e.g. it needs to be linked in to any kernel using cooperative grids.

**Comment by isuruf** [comment]
> This is needed for JIT compilers such as Numba, which compile and link code at runtime - e.g. it needs to be linked in to any kernel using cooperative grids.

Then they should go in a separate `cudatoolkit-static` package. Can you please open an issue?

**Comment by jakirkham** [comment]
> > AFAICT that should already be included ( [a307b6c](https://github.com/conda-forge/cudatoolkit-feedstock/commit/a307b6c62708972cb35ab2de48a0d4c65d2a14d9) ).
> 
> Sorry, I was looking at `10.1.243` and it isn't included compared to defaults.

No worries. I think Anthony wasn't able to find them. Maybe they were in a non-standard location? Please feel free to file an issue for this.

> > Looking at defaults package they actually have these too.
> 
> They aren't in `cudatoolkit-10.1.243-h6bb024c_0`

Was looking at the 11.0 packages.

**Comment by jakirkham** [comment]
Anyways I wouldn't get too hung up on what `defaults` packages of `cudatoolkit` do or don't include certain files as only the latest version got updates. While there was discussion about maintaining old versions ( https://github.com/AnacondaRecipes/cudatoolkit-feedstock/issues/8 ), we never came up with a good system to do that. This something we are trying to remedy in this feedstock ( https://github.com/conda-forge/cudatoolkit-feedstock/issues/8 ).

**Comment by isuruf** [comment]
> Anyways I wouldn't get too hung up on what defaults packages of cudatoolkit do or don't include certain files as only the latest version got updates.

That's just wrong. If we don't include a file that `defaults` does, then we'll be breaking users who rely on it.

**Comment by jakirkham** [comment]
I don't know that I would agree with that conclusion. In part as it seems to not be recognizing the issues identified above with updating the `defaults` package itself. The goal of this package is to replace the `defaults` package (at least as far as conda-forge is concerned). In this way it is no different than any other package we produce. As long as we are making sure we are not breaking users of that package here and responding to issues that come up here (both of which we have done in this thread 😉), I think things should be fine 🙂

**Comment by isuruf** [comment]
I see a bunch of text but no answer to my question. `defaults` package has `libnvjpeg.so*`, but conda-forge package doesn't. How can thing be fine if the conda-forge package doesn't have that?

**Comment by jaimergp** [comment]
So we need that check to back to `10.1`?

https://github.com/conda-forge/cudatoolkit-feedstock/commit/a307b6c62708972cb35ab2de48a0d4c65d2a14d9#diff-e31b1daa1804bd5a01e278db35319af6250411c016e564782592544b013154c5R97

**Comment by gmarkall** [comment]
> > This is needed for JIT compilers such as Numba, which compile and link code at runtime - e.g. it needs to be linked in to any kernel using cooperative grids.
> 
> Then they should go in a separate `cudatoolkit-static` package. Can you please open an issue?

https://github.com/conda-forge/cudatoolkit-feedstock/issues/27

**Comment by mike-wendt** [comment]
@jakirkham I'd say we should use the versioning of the `libcudart.so` not the a version like `11.1.1`

All of the CTK pkgs in `anaconda` & `nvidia` channels use this approach from the old [recipe](https://github.com/AnacondaRecipes/cudatoolkit-feedstock/blob/f5573bbb8048530b86ee03855d54b31126574599/meta.yaml#L3-L4)

This would prevent the issues of `11.1.1` not being pulled instead of `11.1.74` for example with the new CTK 11.1 in #26

**Comment by kkraus14** [comment]
+1 to standardizing on the CUDA release version at least for now.

If we decide to split the packages in the future we could consider releasing them based on their individual versions and make `cudatoolkit` a metapackage that versions things appropriately.

**Comment by jakirkham** [comment]
> I see no answer to my question. `defaults` package has `libnvjpeg.so*`, but conda-forge package doesn't. How can things be fine if the conda-forge package doesn't have that?

It was already answered here ( https://github.com/conda-forge/cudatoolkit-feedstock/issues/15#issuecomment-722898530 ) with a follow-up here ( https://github.com/conda-forge/cudatoolkit-feedstock/issues/15#issuecomment-722927224 ). Again would request that you please file a new issue. I think this conversation is getting hard to track already because it is getting intermingled with other discussion. A new issue would allow for clear discussion around this topic. Thanks in advance! 😄

Also would ask that you please include context about where this library is being used. If it is something that a package is actively linking against, agree we should prioritize it. However I'm not aware of any such use case myself and don't recall this coming up (though could have missed that for the reasons already noted above 😉).

**Comment by isuruf** [comment]
> Also would ask that you please include context about where this library is being used. If it is something that a package is actively linking against, agree we should prioritize it. However I'm not aware of any such use case myself and don't recall this coming up (though could have missed that for the reasons already noted above wink).

I have no idea where this library is used, but if it is added in `defaults` and it's a shared library it's better to err on the side of including it. 10.1 doesn't have it. If your argument is that "I don't know anybody using it, so let's remove it", I won't waste my time anymore.

**Comment by jakirkham** [comment]
> @jakirkham I'd say we should use the versioning of the `libcudart.so` not the a version like `11.1.1`
> 
> All of the CTK pkgs in `anaconda` & `nvidia` channels use this approach from the old [recipe](https://github.com/AnacondaRecipes/cudatoolkit-feedstock/blob/f5573bbb8048530b86ee03855d54b31126574599/meta.yaml#L3-L4)
> 
> This would prevent the issues of `11.1.1` not being pulled instead of `11.1.74` for example with the new CTK 11.1 in #26

I don't have strong thoughts here. Am happy for us to do whatever makes sense.

@scopatz, WDYT?

**Comment by jakirkham** [comment]
> > Also would ask that you please include context about where this library is being used. If it is something that a package is actively linking against, agree we should prioritize it. However I'm not aware of any such use case myself and don't recall this coming up (though could have missed that for the reasons already noted above wink).
> 
> I have no idea where this library is used, but if it is added in `defaults` and it's a shared library it's better to err on the side of including it. 10.1 doesn't have it. If your argument is that "I don't know anybody using it, so let's remove it", I won't waste my time anymore.

I'm not saying let's just remove it. I'm saying Anthony tried to get this one into the package, but wasn't able to (again please read my previous comment). If someone has time to dig in here, that would be most welcome. Unfortunately if you are asking me to fix it, I'm not able to look into that currently. If the general consensus is no one has time to look (which is fine), we should have an issue to track it. Please feel free to raise that issue 🙂

**Comment by isuruf** [comment]
I opened #29 and #30. I'm not asking you to fix it. I'm just saying that we should investigate this before moving these to `main` label.

**Comment by jakirkham** [comment]
Thank you for doing that Isuru! Also thank you isolating these issues. Sorry if my previous comments came up off as unappreciative, I do appreciate the work you are doing here 🙂

**Comment by jakirkham** [comment]
So it looks like those 2 issues got fixed. Is there anything else we should do before publishing to `conda-forge/main`?

**Comment by kkraus14** [comment]
> So it looks like those 2 issues got fixed. Is there anything else we should do before publishing to `conda-forge/main`?

I think we need to handle the missing cupti libs still.

**Comment by jakirkham** [comment]
Are these in the Anaconda packages? Do we have a use case for them? Asking both to figure out if missing them will cause issues and to see if we have anyway to confirm we got them packaged correctly (certainly a consideration for Windows).

**Comment by jaimergp** [comment]
These are the contents of all `defaults` packages for Linux and Windows. None of them contain `*cupti*` stuff.


<details>

<summary>linux-64 / cudatoolkit-9.2-0.tar.bz2</summary>

```
.
├── info
│   ├── about.json
│   ├── files
│   ├── git
│   ├── hash_input.json
│   ├── index.json
│   ├── paths.json
│   ├── recipe
│   │   ├── bld.bat
│   │   ├── build.sh
│   │   ├── conda_build_config.yaml
│   │   ├── meta.yaml
│   │   ├── meta.yaml.template
│   │   └── run_test.py
│   └── test
│       ├── run_test.py
│       └── test_time_dependencies.json
└── lib
    ├── cudatoolkit_config.yaml
    ├── libcublas.so
    ├── libcublas.so.9.2
    ├── libcublas.so.9.2.148
    ├── libcudart.so
    ├── libcudart.so.9.2
    ├── libcudart.so.9.2.148
    ├── libcufft.so
    ├── libcufft.so.9.2
    ├── libcufft.so.9.2.148
    ├── libcufftw.so
    ├── libcufftw.so.9.2
    ├── libcufftw.so.9.2.148
    ├── libcurand.so
    ├── libcurand.so.9.2
    ├── libcurand.so.9.2.148
    ├── libcusolver.so
    ├── libcusolver.so.9.2
    ├── libcusolver.so.9.2.148
    ├── libcusparse.so
    ├── libcusparse.so.9.2
    ├── libcusparse.so.9.2.148
    ├── libdevice.10.bc
    ├── libnppc.so
    ├── libnppc.so.9.2
    ├── libnppc.so.9.2.148
    ├── libnppial.so
    ├── libnppial.so.9.2
    ├── libnppial.so.9.2.148
    ├── libnppicc.so
    ├── libnppicc.so.9.2
    ├── libnppicc.so.9.2.148
    ├── libnppicom.so
    ├── libnppicom.so.9.2
    ├── libnppicom.so.9.2.148
    ├── libnppidei.so
    ├── libnppidei.so.9.2
    ├── libnppidei.so.9.2.148
    ├── libnppif.so
    ├── libnppif.so.9.2
    ├── libnppif.so.9.2.148
    ├── libnppig.so
    ├── libnppig.so.9.2
    ├── libnppig.so.9.2.148
    ├── libnppim.so
    ├── libnppim.so.9.2
    ├── libnppim.so.9.2.148
    ├── libnppist.so
    ├── libnppist.so.9.2
    ├── libnppist.so.9.2.148
    ├── libnppisu.so
    ├── libnppisu.so.9.2
    ├── libnppisu.so.9.2.148
    ├── libnppitc.so
    ├── libnppitc.so.9.2
    ├── libnppitc.so.9.2.148
    ├── libnpps.so
    ├── libnpps.so.9.2
    ├── libnpps.so.9.2.148
    ├── libnvToolsExt.so
    ├── libnvToolsExt.so.1
    ├── libnvToolsExt.so.1.0.0
    ├── libnvblas.so
    ├── libnvblas.so.9.2
    ├── libnvblas.so.9.2.148
    ├── libnvgraph.so
    ├── libnvgraph.so.9.2
    ├── libnvgraph.so.9.2.148
    ├── libnvrtc-builtins.so
    ├── libnvrtc-builtins.so.9.2
    ├── libnvrtc-builtins.so.9.2.148
    ├── libnvrtc.so
    ├── libnvrtc.so.9.2
    ├── libnvrtc.so.9.2.148
    ├── libnvvm.so
    ├── libnvvm.so.3
    └── libnvvm.so.3.2.0

4 directories, 91 files

```

</details>

<details>

<summary>win-64 / cudatoolkit-10.0.130-0.tar.bz2</summary>

```
.
├── DLLs
│   ├── libdevice.10.bc
│   └── nvvm64_33_0.dll
├── Library
│   └── bin
│       ├── cublas64_100.dll
│       ├── cudart64_100.dll
│       ├── cudatoolkit_config.yaml
│       ├── cufft64_100.dll
│       ├── cufftw64_100.dll
│       ├── curand64_100.dll
│       ├── cusolver64_100.dll
│       ├── cusparse64_100.dll
│       ├── libdevice.10.bc
│       ├── nppc64_100.dll
│       ├── nppial64_100.dll
│       ├── nppicc64_100.dll
│       ├── nppicom64_100.dll
│       ├── nppidei64_100.dll
│       ├── nppif64_100.dll
│       ├── nppig64_100.dll
│       ├── nppim64_100.dll
│       ├── nppist64_100.dll
│       ├── nppisu64_100.dll
│       ├── nppitc64_100.dll
│       ├── npps64_100.dll
│       ├── nvToolsExt64_1.dll
│       ├── nvblas64_100.dll
│       ├── nvgraph64_100.dll
│       ├── nvrtc-builtins64_100.dll
│       ├── nvrtc64_100_0.dll
│       └── nvvm64_33_0.dll
└── info
    ├── about.json
    ├── files
    ├── git
    ├── hash_input.json
    ├── index.json
    ├── paths.json
    ├── recipe
    │   ├── bld.bat
    │   ├── build.sh
    │   ├── conda_build_config.yaml
    │   ├── meta.yaml
    │   ├── meta.yaml.template
    │   └── run_test.py
    └── test
        ├── run_test.py
        └── test_time_dependencies.json

6 directories, 43 files

```

</details>

<details>

<summary>linux-64 / cudatoolkit-10.0.130-0.tar.bz2</summary>

```
.
├── info
│   ├── about.json
│   ├── files
│   ├── git
│   ├── hash_input.json
│   ├── index.json
│   ├── paths.json
│   ├── recipe
│   │   ├── bld.bat
│   │   ├── build.sh
│   │   ├── conda_build_config.yaml
│   │   ├── meta.yaml
│   │   ├── meta.yaml.template
│   │   └── run_test.py
│   └── test
│       ├── run_test.py
│       └── test_time_dependencies.json
└── lib
    ├── cudatoolkit_config.yaml
    ├── libcublas.so
    ├── libcublas.so.10.0
    ├── libcublas.so.10.0.130
    ├── libcudart.so
    ├── libcudart.so.10.0
    ├── libcudart.so.10.0.130
    ├── libcufft.so
    ├── libcufft.so.10.0
    ├── libcufft.so.10.0.145
    ├── libcufftw.so
    ├── libcufftw.so.10.0
    ├── libcufftw.so.10.0.145
    ├── libcurand.so
    ├── libcurand.so.10.0
    ├── libcurand.so.10.0.130
    ├── libcusolver.so
    ├── libcusolver.so.10.0
    ├── libcusolver.so.10.0.130
    ├── libcusparse.so
    ├── libcusparse.so.10.0
    ├── libcusparse.so.10.0.130
    ├── libdevice.10.bc
    ├── libnppc.so
    ├── libnppc.so.10.0
    ├── libnppc.so.10.0.130
    ├── libnppial.so
    ├── libnppial.so.10.0
    ├── libnppial.so.10.0.130
    ├── libnppicc.so
    ├── libnppicc.so.10.0
    ├── libnppicc.so.10.0.130
    ├── libnppicom.so
    ├── libnppicom.so.10.0
    ├── libnppicom.so.10.0.130
    ├── libnppidei.so
    ├── libnppidei.so.10.0
    ├── libnppidei.so.10.0.130
    ├── libnppif.so
    ├── libnppif.so.10.0
    ├── libnppif.so.10.0.130
    ├── libnppig.so
    ├── libnppig.so.10.0
    ├── libnppig.so.10.0.130
    ├── libnppim.so
    ├── libnppim.so.10.0
    ├── libnppim.so.10.0.130
    ├── libnppist.so
    ├── libnppist.so.10.0
    ├── libnppist.so.10.0.130
    ├── libnppisu.so
    ├── libnppisu.so.10.0
    ├── libnppisu.so.10.0.130
    ├── libnppitc.so
    ├── libnppitc.so.10.0
    ├── libnppitc.so.10.0.130
    ├── libnpps.so
    ├── libnpps.so.10.0
    ├── libnpps.so.10.0.130
    ├── libnvToolsExt.so
    ├── libnvToolsExt.so.1
    ├── libnvToolsExt.so.1.0.0
    ├── libnvblas.so
    ├── libnvblas.so.10.0
    ├── libnvblas.so.10.0.130
    ├── libnvgraph.so
    ├── libnvgraph.so.10.0
    ├── libnvgraph.so.10.0.130
    ├── libnvrtc-builtins.so
    ├── libnvrtc-builtins.so.10.0
    ├── libnvrtc-builtins.so.10.0.130
    ├── libnvrtc.so
    ├── libnvrtc.so.10.0
    ├── libnvrtc.so.10.0.130
    ├── libnvvm.so
    ├── libnvvm.so.3
    └── libnvvm.so.3.3.0

4 directories, 91 files

```

</details>

<details>

<summary>linux-64 / cudatoolkit-10.1.168-0.tar.bz2</summary>

```
.
├── info
│   ├── about.json
│   ├── files
│   ├── git
│   ├── hash_input.json
│   ├── index.json
│   ├── paths.json
│   ├── recipe
│   │   ├── bld.bat
│   │   ├── build.sh
│   │   ├── conda_build_config.yaml
│   │   ├── meta.yaml
│   │   ├── meta.yaml.template
│   │   └── run_test.py
│   └── test
│       ├── run_test.py
│       └── test_time_dependencies.json
└── lib
    ├── cudatoolkit_config.yaml
    ├── libcublas.so
    ├── libcublas.so.10
    ├── libcublas.so.10.2.0.168
    ├── libcublasLt.so
    ├── libcublasLt.so.10
    ├── libcublasLt.so.10.2.0.168
    ├── libcudart.so
    ├── libcudart.so.10.1
    ├── libcudart.so.10.1.168
    ├── libcufft.so
    ├── libcufft.so.10
    ├── libcufft.so.10.1.168
    ├── libcufftw.so
    ├── libcufftw.so.10
    ├── libcufftw.so.10.1.168
    ├── libcurand.so
    ├── libcurand.so.10
    ├── libcurand.so.10.1.168
    ├── libcusolver.so
    ├── libcusolver.so.10
    ├── libcusolver.so.10.1.168
    ├── libcusparse.so
    ├── libcusparse.so.10
    ├── libcusparse.so.10.1.168
    ├── libdevice.10.bc
    ├── libnppc.so
    ├── libnppc.so.10
    ├── libnppc.so.10.1.168
    ├── libnppial.so
    ├── libnppial.so.10
    ├── libnppial.so.10.1.168
    ├── libnppicc.so
    ├── libnppicc.so.10
    ├── libnppicc.so.10.1.168
    ├── libnppicom.so
    ├── libnppicom.so.10
    ├── libnppicom.so.10.1.168
    ├── libnppidei.so
    ├── libnppidei.so.10
    ├── libnppidei.so.10.1.168
    ├── libnppif.so
    ├── libnppif.so.10
    ├── libnppif.so.10.1.168
    ├── libnppig.so
    ├── libnppig.so.10
    ├── libnppig.so.10.1.168
    ├── libnppim.so
    ├── libnppim.so.10
    ├── libnppim.so.10.1.168
    ├── libnppist.so
    ├── libnppist.so.10
    ├── libnppist.so.10.1.168
    ├── libnppisu.so
    ├── libnppisu.so.10
    ├── libnppisu.so.10.1.168
    ├── libnppitc.so
    ├── libnppitc.so.10
    ├── libnppitc.so.10.1.168
    ├── libnpps.so
    ├── libnpps.so.10
    ├── libnpps.so.10.1.168
    ├── libnvToolsExt.so
    ├── libnvToolsExt.so.1
    ├── libnvToolsExt.so.1.0.0
    ├── libnvblas.so
    ├── libnvblas.so.10
    ├── libnvblas.so.10.2.0.168
    ├── libnvgraph.so
    ├── libnvgraph.so.10
    ├── libnvgraph.so.10.1.168
    ├── libnvjpeg.so
    ├── libnvjpeg.so.10
    ├── libnvjpeg.so.10.1.168
    ├── libnvrtc-builtins.so
    ├── libnvrtc-builtins.so.10.1
    ├── libnvrtc-builtins.so.10.1.168
    ├── libnvrtc.so
    ├── libnvrtc.so.10.1
    ├── libnvrtc.so.10.1.168
    ├── libnvvm.so
    ├── libnvvm.so.3
    └── libnvvm.so.3.3.0

4 directories, 97 files

```

</details>

<details>

<summary>linux-64 / cudatoolkit-10.1.243-h6bb024c_0.tar.bz2</summary>

```
.
├── info
│   ├── about.json
│   ├── files
│   ├── git
│   ├── hash_input.json
│   ├── index.json
│   ├── paths.json
│   ├── recipe
│   │   ├── LICENSE
│   │   ├── bld.bat
│   │   ├── build.py
│   │   ├── build.sh
│   │   ├── conda_build_config.yaml
│   │   ├── meta.yaml
│   │   ├── meta.yaml.template
│   │   └── run_test.py
│   └── test
│       ├── run_test.py
│       └── test_time_dependencies.json
└── lib
    ├── cudatoolkit_config.yaml
    ├── libcublas.so
    ├── libcublas.so.10
    ├── libcublas.so.10.2.1.243
    ├── libcublasLt.so
    ├── libcublasLt.so.10
    ├── libcublasLt.so.10.2.1.243
    ├── libcudart.so
    ├── libcudart.so.10.1
    ├── libcudart.so.10.1.243
    ├── libcufft.so
    ├── libcufft.so.10
    ├── libcufft.so.10.1.1.243
    ├── libcufftw.so
    ├── libcufftw.so.10
    ├── libcufftw.so.10.1.1.243
    ├── libcurand.so
    ├── libcurand.so.10
    ├── libcurand.so.10.1.1.243
    ├── libcusolver.so
    ├── libcusolver.so.10
    ├── libcusolver.so.10.2.0.243
    ├── libcusparse.so
    ├── libcusparse.so.10
    ├── libcusparse.so.10.3.0.243
    ├── libdevice.10.bc
    ├── libnppc.so
    ├── libnppc.so.10
    ├── libnppc.so.10.2.0.243
    ├── libnppial.so
    ├── libnppial.so.10
    ├── libnppial.so.10.2.0.243
    ├── libnppicc.so
    ├── libnppicc.so.10
    ├── libnppicc.so.10.2.0.243
    ├── libnppicom.so
    ├── libnppicom.so.10
    ├── libnppicom.so.10.2.0.243
    ├── libnppidei.so
    ├── libnppidei.so.10
    ├── libnppidei.so.10.2.0.243
    ├── libnppif.so
    ├── libnppif.so.10
    ├── libnppif.so.10.2.0.243
    ├── libnppig.so
    ├── libnppig.so.10
    ├── libnppig.so.10.2.0.243
    ├── libnppim.so
    ├── libnppim.so.10
    ├── libnppim.so.10.2.0.243
    ├── libnppist.so
    ├── libnppist.so.10
    ├── libnppist.so.10.2.0.243
    ├── libnppisu.so
    ├── libnppisu.so.10
    ├── libnppisu.so.10.2.0.243
    ├── libnppitc.so
    ├── libnppitc.so.10
    ├── libnppitc.so.10.2.0.243
    ├── libnpps.so
    ├── libnpps.so.10
    ├── libnpps.so.10.2.0.243
    ├── libnvToolsExt.so
    ├── libnvToolsExt.so.1
    ├── libnvToolsExt.so.1.0.0
    ├── libnvblas.so
    ├── libnvblas.so.10
    ├── libnvblas.so.10.2.1.243
    ├── libnvgraph.so
    ├── libnvgraph.so.10
    ├── libnvgraph.so.10.1.243
    ├── libnvjpeg.so
    ├── libnvjpeg.so.10
    ├── libnvjpeg.so.10.3.0.243
    ├── libnvrtc-builtins.so
    ├── libnvrtc-builtins.so.10.1
    ├── libnvrtc-builtins.so.10.1.243
    ├── libnvrtc.so
    ├── libnvrtc.so.10.1
    ├── libnvrtc.so.10.1.243
    ├── libnvvm.so
    ├── libnvvm.so.3
    └── libnvvm.so.3.3.0

4 directories, 99 files

```

</details>

<details>

<summary>win-64 / cudatoolkit-10.1.243-h74a9793_0.tar.bz2</summary>

```
.
├── DLLs
│   ├── libdevice.10.bc
│   └── nvvm64_33_0.dll
├── Library
│   └── bin
│       ├── cublas64_10.dll
│       ├── cublasLt64_10.dll
│       ├── cudart64_101.dll
│       ├── cudatoolkit_config.yaml
│       ├── cufft64_10.dll
│       ├── cufftw64_10.dll
│       ├── curand64_10.dll
│       ├── cusolver64_10.dll
│       ├── cusparse64_10.dll
│       ├── libdevice.10.bc
│       ├── nppc64_10.dll
│       ├── nppial64_10.dll
│       ├── nppicc64_10.dll
│       ├── nppicom64_10.dll
│       ├── nppidei64_10.dll
│       ├── nppif64_10.dll
│       ├── nppig64_10.dll
│       ├── nppim64_10.dll
│       ├── nppist64_10.dll
│       ├── nppisu64_10.dll
│       ├── nppitc64_10.dll
│       ├── npps64_10.dll
│       ├── nvToolsExt64_1.dll
│       ├── nvblas64_10.dll
│       ├── nvgraph64_10.dll
│       ├── nvrtc-builtins64_101.dll
│       ├── nvrtc64_101_0.dll
│       └── nvvm64_33_0.dll
└── info
    ├── about.json
    ├── files
    ├── git
    ├── hash_input.json
    ├── index.json
    ├── paths.json
    ├── recipe
    │   ├── LICENSE
    │   ├── bld.bat
    │   ├── build.py
    │   ├── build.sh
    │   ├── conda_build_config.yaml
    │   ├── meta.yaml
    │   ├── meta.yaml.template
    │   └── run_test.py
    └── test
        ├── run_test.py
        └── test_time_dependencies.json

6 directories, 46 files

```

</details>

<details>

<summary>win-64 / cudatoolkit-10.2.89-h74a9793_1.tar.bz2</summary>

```
.
├── DLLs
│   ├── libdevice.10.bc
│   └── nvvm64_33_0.dll
├── Library
│   └── bin
│       ├── cublas64_10.dll
│       ├── cublasLt64_10.dll
│       ├── cudadevrt.lib
│       ├── cudart64_102.dll
│       ├── cudatoolkit_config.yaml
│       ├── cufft64_10.dll
│       ├── cufftw64_10.dll
│       ├── curand64_10.dll
│       ├── cusolver64_10.dll
│       ├── cusparse64_10.dll
│       ├── libdevice.10.bc
│       ├── nppc64_10.dll
│       ├── nppial64_10.dll
│       ├── nppicc64_10.dll
│       ├── nppicom64_10.dll
│       ├── nppidei64_10.dll
│       ├── nppif64_10.dll
│       ├── nppig64_10.dll
│       ├── nppim64_10.dll
│       ├── nppist64_10.dll
│       ├── nppisu64_10.dll
│       ├── nppitc64_10.dll
│       ├── npps64_10.dll
│       ├── nvToolsExt64_1.dll
│       ├── nvblas64_10.dll
│       ├── nvgraph64_10.dll
│       ├── nvrtc-builtins64_102.dll
│       ├── nvrtc64_102_0.dll
│       └── nvvm64_33_0.dll
└── info
    ├── about.json
    ├── files
    ├── git
    ├── hash_input.json
    ├── index.json
    ├── paths.json
    ├── recipe
    │   ├── LICENSE
    │   ├── bld.bat
    │   ├── build.py
    │   ├── build.sh
    │   ├── conda_build_config.yaml
    │   ├── meta.yaml
    │   ├── meta.yaml.template
    │   ├── recipe_log.txt
    │   └── run_test.py
    └── test
        ├── run_test.py
        └── test_time_dependencies.json

6 directories, 48 files

```

</details>

<details>

<summary>linux-64 / cudatoolkit-10.2.89-hfd86e86_1.tar.bz2</summary>

```
.
├── info
│   ├── about.json
│   ├── files
│   ├── git
│   ├── hash_input.json
│   ├── index.json
│   ├── paths.json
│   ├── recipe
│   │   ├── LICENSE
│   │   ├── bld.bat
│   │   ├── build.py
│   │   ├── build.sh
│   │   ├── conda_build_config.yaml
│   │   ├── meta.yaml
│   │   ├── meta.yaml.template
│   │   ├── recipe_log.txt
│   │   └── run_test.py
│   └── test
│       ├── run_test.py
│       └── test_time_dependencies.json
└── lib
    ├── cudatoolkit_config.yaml
    ├── libcublas.so
    ├── libcublas.so.10
    ├── libcublas.so.10.2.2.89
    ├── libcublasLt.so
    ├── libcublasLt.so.10
    ├── libcublasLt.so.10.2.2.89
    ├── libcudadevrt.a
    ├── libcudart.so
    ├── libcudart.so.10.2
    ├── libcudart.so.10.2.89
    ├── libcufft.so
    ├── libcufft.so.10
    ├── libcufft.so.10.1.2.89
    ├── libcufftw.so
    ├── libcufftw.so.10
    ├── libcufftw.so.10.1.2.89
    ├── libcurand.so
    ├── libcurand.so.10
    ├── libcurand.so.10.1.2.89
    ├── libcusolver.so
    ├── libcusolver.so.10
    ├── libcusolver.so.10.3.0.89
    ├── libcusparse.so
    ├── libcusparse.so.10
    ├── libcusparse.so.10.3.1.89
    ├── libdevice.10.bc
    ├── libnppc.so
    ├── libnppc.so.10
    ├── libnppc.so.10.2.1.89
    ├── libnppial.so
    ├── libnppial.so.10
    ├── libnppial.so.10.2.1.89
    ├── libnppicc.so
    ├── libnppicc.so.10
    ├── libnppicc.so.10.2.1.89
    ├── libnppicom.so
    ├── libnppicom.so.10
    ├── libnppicom.so.10.2.1.89
    ├── libnppidei.so
    ├── libnppidei.so.10
    ├── libnppidei.so.10.2.1.89
    ├── libnppif.so
    ├── libnppif.so.10
    ├── libnppif.so.10.2.1.89
    ├── libnppig.so
    ├── libnppig.so.10
    ├── libnppig.so.10.2.1.89
    ├── libnppim.so
    ├── libnppim.so.10
    ├── libnppim.so.10.2.1.89
    ├── libnppist.so
    ├── libnppist.so.10
    ├── libnppist.so.10.2.1.89
    ├── libnppisu.so
    ├── libnppisu.so.10
    ├── libnppisu.so.10.2.1.89
    ├── libnppitc.so
    ├── libnppitc.so.10
    ├── libnppitc.so.10.2.1.89
    ├── libnpps.so
    ├── libnpps.so.10
    ├── libnpps.so.10.2.1.89
    ├── libnvToolsExt.so
    ├── libnvToolsExt.so.1
    ├── libnvToolsExt.so.1.0.0
    ├── libnvblas.so
    ├── libnvblas.so.10
    ├── libnvblas.so.10.2.2.89
    ├── libnvgraph.so
    ├── libnvgraph.so.10
    ├── libnvgraph.so.10.2.89
    ├── libnvjpeg.so
    ├── libnvjpeg.so.10
    ├── libnvjpeg.so.10.3.1.89
    ├── libnvrtc-builtins.so
    ├── libnvrtc-builtins.so.10.2
    ├── libnvrtc-builtins.so.10.2.89
    ├── libnvrtc.so
    ├── libnvrtc.so.10.2
    ├── libnvrtc.so.10.2.89
    ├── libnvvm.so
    ├── libnvvm.so.3
    └── libnvvm.so.3.3.0

4 directories, 101 files

```

</details>

<details>

<summary>linux-64 / cudatoolkit-11.0.221-h6bb024c_0.tar.bz2</summary>

```
.
├── bin
├── info
│   ├── about.json
│   ├── files
│   ├── git
│   ├── hash_input.json
│   ├── index.json
│   ├── licenses
│   │   └── NVIDIA_EULA
│   ├── paths.json
│   ├── recipe
│   │   ├── LICENSE
│   │   ├── NVIDIA_EULA
│   │   ├── bld.bat
│   │   ├── build.py
│   │   ├── build.sh
│   │   ├── conda_build_config.yaml
│   │   ├── meta.yaml
│   │   ├── meta.yaml.template
│   │   ├── post-link.sh
│   │   └── run_test.py
│   └── test
│       ├── run_test.py
│       └── test_time_dependencies.json
└── lib
    ├── cudatoolkit_config.yaml
    ├── libaccinj64.so
    ├── libaccinj64.so.11.0
    ├── libaccinj64.so.11.0.221
    ├── libcublas.so
    ├── libcublas.so.11
    ├── libcublas.so.11.2.0.252
    ├── libcublasLt.so
    ├── libcublasLt.so.11
    ├── libcublasLt.so.11.2.0.252
    ├── libcudadevrt.a
    ├── libcudart.so
    ├── libcudart.so.11.0
    ├── libcudart.so.11.0.221
    ├── libcufft.so
    ├── libcufft.so.10
    ├── libcufft.so.10.2.1.245
    ├── libcufftw.so
    ├── libcufftw.so.10
    ├── libcufftw.so.10.2.1.245
    ├── libcuinj64.so
    ├── libcuinj64.so.11.0
    ├── libcuinj64.so.11.0.221
    ├── libcurand.so
    ├── libcurand.so.10
    ├── libcurand.so.10.2.1.245
    ├── libcusolver.so
    ├── libcusolver.so.10
    ├── libcusolver.so.10.6.0.245
    ├── libcusolverMg.so
    ├── libcusolverMg.so.10
    ├── libcusolverMg.so.10.6.0.245
    ├── libcusparse.so
    ├── libcusparse.so.11
    ├── libcusparse.so.11.1.1.245
    ├── libdevice.10.bc
    ├── libnppc.so
    ├── libnppc.so.11
    ├── libnppc.so.11.1.0.245
    ├── libnppial.so
    ├── libnppial.so.11
    ├── libnppial.so.11.1.0.245
    ├── libnppicc.so
    ├── libnppicc.so.11
    ├── libnppicc.so.11.1.0.245
    ├── libnppidei.so
    ├── libnppidei.so.11
    ├── libnppidei.so.11.1.0.245
    ├── libnppif.so
    ├── libnppif.so.11
    ├── libnppif.so.11.1.0.245
    ├── libnppig.so
    ├── libnppig.so.11
    ├── libnppig.so.11.1.0.245
    ├── libnppim.so
    ├── libnppim.so.11
    ├── libnppim.so.11.1.0.245
    ├── libnppist.so
    ├── libnppist.so.11
    ├── libnppist.so.11.1.0.245
    ├── libnppisu.so
    ├── libnppisu.so.11
    ├── libnppisu.so.11.1.0.245
    ├── libnppitc.so
    ├── libnppitc.so.11
    ├── libnppitc.so.11.1.0.245
    ├── libnpps.so
    ├── libnpps.so.11
    ├── libnpps.so.11.1.0.245
    ├── libnvToolsExt.so
    ├── libnvToolsExt.so.1
    ├── libnvToolsExt.so.1.0.0
    ├── libnvblas.so
    ├── libnvblas.so.11
    ├── libnvblas.so.11.2.0.252
    ├── libnvjpeg.so
    ├── libnvjpeg.so.11
    ├── libnvjpeg.so.11.1.1.245
    ├── libnvrtc-builtins.so
    ├── libnvrtc-builtins.so.11.0
    ├── libnvrtc-builtins.so.11.0.221
    ├── libnvrtc.so
    ├── libnvrtc.so.11.0
    ├── libnvrtc.so.11.0.221
    ├── libnvvm.so
    ├── libnvvm.so.3
    └── libnvvm.so.3.3.0

6 directories, 106 files

```

</details>

<details>

<summary>win-64 / cudatoolkit-11.0.221-h74a9793_0.tar.bz2</summary>

```
.
├── DLLs
│   ├── libdevice.10.bc
│   └── nvvm64_33_0.dll
├── Library
│   └── bin
│       ├── cublas64_11.dll
│       ├── cublasLt64_11.dll
│       ├── cudadevrt.lib
│       ├── cudart64_110.dll
│       ├── cudatoolkit_config.yaml
│       ├── cufft64_10.dll
│       ├── cufftw64_10.dll
│       ├── curand64_10.dll
│       ├── cusolver64_10.dll
│       ├── cusolverMg64_10.dll
│       ├── cusparse64_11.dll
│       ├── libdevice.10.bc
│       ├── nppc64_11.dll
│       ├── nppial64_11.dll
│       ├── nppicc64_11.dll
│       ├── nppidei64_11.dll
│       ├── nppif64_11.dll
│       ├── nppig64_11.dll
│       ├── nppim64_11.dll
│       ├── nppist64_11.dll
│       ├── nppisu64_11.dll
│       ├── nppitc64_11.dll
│       ├── npps64_11.dll
│       ├── nvToolsExt64_1.dll
│       ├── nvblas64_11.dll
│       ├── nvjpeg64_11.dll
│       ├── nvrtc-builtins64_110.dll
│       ├── nvrtc64_110_0.dll
│       └── nvvm64_33_0.dll
└── info
    ├── about.json
    ├── files
    ├── git
    ├── hash_input.json
    ├── index.json
    ├── licenses
    │   └── NVIDIA_EULA
    ├── paths.json
    ├── recipe
    │   ├── LICENSE
    │   ├── NVIDIA_EULA
    │   ├── bld.bat
    │   ├── build.py
    │   ├── build.sh
    │   ├── conda_build_config.yaml
    │   ├── meta.yaml
    │   ├── meta.yaml.template
    │   ├── post-link.sh
    │   └── run_test.py
    └── test
        ├── run_test.py
        └── test_time_dependencies.json

7 directories, 50 files
```

</details>

**Comment by ax3l** [comment]
I have a general question on shipping cudatoolkit versions and practical workflows.

My users use user-level package managers on systems where they cannot quickly install system-level software. Assume the following situtation: they are on a shared machine/cluster that has drivers pre-installed for, e.g. `CUDA 10.1.243`. If they try to pull now a CUDA package from conda-forge, which builds against the latest CUDA toolkit release by default, they will not be able to use this since the drivers are too old for the cuda we built against, right?

Should feedstock maintainers thus try to build always against the oldest cuda toolkit they can build against? Or build against multiple versions?

**Comment by kkraus14** [comment]
> If they try to pull now a CUDA package from conda-forge, which builds against the latest CUDA toolkit release by default, they will not be able to use this since the drivers are too old for the cuda we built against, right?

There's a `__cuda` virtual package in conda which is the newest supported version of the toolkit that the driver on the system supports. If they try to install a newer toolkit than what their driver supports it will error.

**Comment by jakirkham** [comment]
Also the way things are configured now is we build against a matrix of different CUDA versions. That way users can download the packages supporting the latest version of CUDA their machine supports. There shouldn't be anything special feedstock maintainers need to do to enable this beyond adding `{{ compiler("cuda") }}` to `build`. Here's an example ( https://github.com/conda-forge/ucx-split-feedstock/pull/14 ) (though ignore the `conda_build_config.yaml` bit as that lives in `conda-forge-pinning` now).

Would add the `__cuda` virtual package is something Conda recently started using with the solver ( https://github.com/conda/conda/pull/10057 ). So please make sure your users have Conda 4.8.4+ to take advantage of this feature. If only an older Conda is available, users can check the `__cuda` version using `conda info` (requires Conda 4.7.0+) ( https://github.com/conda/conda/pull/8267 ) and manually constrain the `cudatoolkit` package in installs to have the same effect.

**Comment by isuruf** [comment]
re: cupti

> Are these in the Anaconda packages? 

They aren't. So it's okay to not wait for that.

**Comment by jakirkham** [comment]
Jaime added PR ( https://github.com/conda-forge/cudatoolkit-feedstock/pull/43 ) (just now merged it), which includes `cupti`. So that should be good now.

IDK what we want to do about the static libraries ( https://github.com/conda-forge/cudatoolkit-feedstock/issues/27 ). Anaconda currently packages those. Though that technically goes against [CFEP-18]( https://github.com/conda-forge/cfep/blob/70f80e245a3497be35cbcf03df91d8b53b4e7931/cfep-18.md ). So not sure how we want to handle this.

Sounds like there are no other blockers?

**Comment by scopatz** [comment]
We make other exceptions for staticlibs for base level packages, so it is probably ok to do so here too, esp since IIUC the static libs are used at runtime by downstream packages, like numba.

**Comment by leofang** [comment]
> We make other exceptions for staticlibs for base level packages, so it is probably ok to do so here too, esp since IIUC the static libs are used at runtime by downstream packages, like numba.

That is exactly what I said in #27 🤝

**Comment by isuruf** [comment]
I disagree that it should be an exception. Downstream packages should depend on the static runtime explicitly. However it's not a blocker for publishing these packages on `main` label.

**Comment by scopatz** [comment]
Yeah, I agree, Isuru, if anyone is willing to do that work that would be better. However, that would break compatibility with defaults unless we had a metapackage (called cudatoolkit which installed both).  If no one is interested in splitting the package, I think it is ok proceed.

**Comment by scopatz** [comment]
A package split can always come later, like it does in other packages, so I don't see this as blocking

**Comment by kkraus14** [comment]
+1 to making `cudatoolkit` a metapackage in the future and that's when we should break out the static libraries we've historically packaged.

**Comment by jakirkham** [comment]
Yeah we could make a `libcudatoolkit` package, which contains the shared objects.

**Comment by jakirkham** [comment]
Anyways it sounds like we are good to go ahead and publish the package. Have submitted PR ( https://github.com/conda-forge/cudatoolkit-feedstock/pull/44 ) to do just that. Thanks all for all of your help trying packages, making fixes, pointing out issues, etc.. This has been very helpful. Hope the new package works well for everyone! 😄
